plutus-tx/src/PlutusTx/DataList.hs

```diff
@@ -0,0 +1,59 @@
+{-# LANGUAGE ConstraintKinds    #-}
+{-# LANGUAGE DerivingStrategies #-}
+{-# LANGUAGE OverloadedLists    #-}
+{-# LANGUAGE PatternSynonyms    #-}
+{-# LANGUAGE TemplateHaskell    #-}
+{-# LANGUAGE TypeFamilies       #-}
+{-# LANGUAGE ViewPatterns       #-}
+
+module PlutusTx.DataList (
+
+) where
+
+import PlutusTx.AsData qualified as AsData
+import PlutusTx.IsData qualified as P
+import PlutusTx.Prelude hiding (map)
+
+import GHC.Exts qualified as H
+
+AsData.asData
+  [d|
+    data List a = Nil | Cons a (List a)
+      deriving newtype (P.ToData, P.FromData, P.UnsafeFromData)
+  |]
+
+type DataElem a = (P.UnsafeFromData a, P.ToData a)
+
+instance (DataElem a) => H.IsList (List a) where
+    type Item (List a) = a
+    fromList [] = Nil
```

Comment:
If using the list syntax means calling `fromList` and `toList` all the time, that would be very expensive, since each call of `fromList` and `toList` goes through the whole list. You can verify that by writing a budget test. If that's the case we should just give up on the list syntax.

Reply:
Yes, unfortunately the `fromList` calls are inserted into the program and are present at runtime. And they're likely to be recursive functions, I highly doubt our optimizer is going to be good enough to optimize them away.

Reply:
Hmm, yeah, I guess what we actually want is to define `asData` on `[a]`. How about `asData [d| data List a = List [a] |]`? Then the `toList/fromList` transformations are whatever is implemented in the `pattern List` that the TH produces, right?